### PR TITLE
[Variadic Generics] distinctly diagnose more types of pack expansion errors

### PR DESCRIPTION
**Pull request by sophiapoirier**

### CODE REVIEW

**Comment by sophiapoirier** [comment]
@swift-ci please smoke test

File: include/swift/Basic/SourceLoc.h [hunk 1]
```diff
@@ -157,7 +157,7 @@ class SourceRange {
 /// A half-open character-based source range.
 class CharSourceRange {
   SourceLoc Start;
-  unsigned ByteLength;
+  unsigned ByteLength{};
```
Comment (slavapestov) [inline include/swift/Basic/SourceLoc.h:160]: Why this change?

File: include/swift/AST/DiagnosticsSema.def [hunk 2]
```diff
@@ -5310,8 +5310,8 @@ ERROR(vararg_not_allowed,none,
       "variadic parameter cannot appear outside of a function parameter list",
       ())
 ERROR(expansion_not_allowed,none,
-      "pack expansion %0 cannot appear outside of a function parameter list, "
-      "function result, tuple element or generic argument list", (Type))
+      "pack expansion repeat each %0 can only appear in a function parameter list, "
```
Comment (slavapestov) [inline include/swift/AST/DiagnosticsSema.def:5313]: Instead of writing 'repeat each' in the diagnostic, which looks funny: `repeat each 'T'`, can you create a PackExpansionType and pass it as %0 to the diagnostic? Then it will say `'repeat each T'`

**Comment by sophiapoirier** [comment]
@swift-ci please smoke test

File: test/type/pack_expansion.swift [hunk 3]
```diff
@@ -70,9 +70,12 @@ func packRef<each T>(_: repeat each T) where each T: P {}
...
 // expected-error@+1 {{'each' cannot be applied to non-pack type 'Int'}}
-func invalidPackRef(_: each Int) {}
+func invalidPackRefEachInt(_: each Int) {}
 
-// expected-error@+1 {{pack reference 'T' can only appear in pack expansion or generic requirement}}
+// expected-error@+1 {{pack expansion 'Int' must contain at least one pack reference}}
+func invalidPackRefRepeatInt(_: repeat Int) {}
+
+// expected-error@+1 {{pack reference 'T' requires expansion using keyword 'repeat'}}
 func packRefOutsideExpansion<each T>(_: each T) {}
```
Comment (slavapestov) [inline test/type/pack_expansion.swift:79]: A few more to test:

```
func packRefOutsideExpansion<each T>(_: Array<each T>) {}
func packRefOutsideExpansion<each T>(_: Array<(each T) -> ()>) {}
func packRefOutsideExpansion<each T>(_: (each T)) {}
func packRefOutsideExpansion<each T>(_: each T.Type) {}
```

**Comment by sophiapoirier** [comment]
@swift-ci please smoke test

**Review by hborla — APPROVED** [review]
